File: cses/07_mathematics/3355_sum-of-four-squares/solution.py
import sys
from math import isqrt


def two_squares(m):
    # Tìm a <= b không âm với a*a + b*b == m, hoặc None. Kỹ thuật two-pointer
    # O(sqrt(m)): tăng lo nếu tổng nhỏ hơn m, giảm hi nếu lớn hơn.
    lo = 0
    hi = isqrt(m)
    while lo <= hi:
        s = lo * lo + hi * hi
        if s == m:
            return lo, hi
        if s < m:
            lo += 1
        else:
            hi -= 1
    return None


def three_squares(m):
    # m được giả thiết biểu diễn được thành tổng 3 số chính phương (không thuộc
    # dạng cấm của Legendre). Duyệt c giảm dần, kiểm tra m - c^2 có tách được
    # thành 2 số chính phương không.
    c = isqrt(m)
    while c >= 0:
        r = two_squares(m - c * c)
        if r is not None:
            a, b = r
            return a, b, c
        c -= 1
    return 0, 0, 0


def four_squares(n):
    if n == 0:
        return 0, 0, 0, 0
    # Bước 1 - khử hết thừa số 4: n = 4^k * t, hệ số nhân lại p = 2^k.
    k = 0
    t = n
    while t % 4 == 0:
        t //= 4
        k += 1
    p = 1 << k  # 2^k
    # Bước 2 - dạng cấm duy nhất còn lại khi t không chia hết cho 4 là t % 8 == 7:
    # tách d = 1 để phần dư t-1 luôn là tổng 3 số chính phương; ngược lại d = 0.
    if t % 8 == 7:
        a, b, c = three_squares(t - 1)
        d = 1
    else:
        a, b, c = three_squares(t)
        d = 0
    # Bước 3 - nhân lại nghiệm của t với p để được nghiệm của n.
    return a * p, b * p, c * p, d * p


def main():
    data = sys.stdin.buffer.read().split()
    if not data:
        return
    t = int(data[0])
    out = []
    for i in range(1, t + 1):
        n = int(data[i])
        a, b, c, d = four_squares(n)
        out.append("%d %d %d %d" % (a, b, c, d))
    sys.stdout.write("\n".join(out))
    if out:
        sys.stdout.write("\n")


main()

File: cses/07_mathematics/3355_sum-of-four-squares/solution.cpp
#include <cmath>
#include <iostream>
#include <tuple>

// Căn bậc hai nguyên: làm tròn từ sqrt double rồi hiệu chỉnh để tránh sai số.
int integer_square_root(const int value) {
    int root = static_cast<int>(std::sqrt(static_cast<double>(value)));
    while ((root + 1) * (root + 1) <= value) {
        ++root;
    }
    while (root * root > value) {
        --root;
    }
    return root;
}

// Kiểm tra value có là tổng 2 số chính phương không bằng kỹ thuật two-pointer:
// tăng low nếu tổng nhỏ hơn value, giảm high nếu lớn hơn. O(sqrt(value)).
bool two_squares(const int value, int& first, int& second) {
    int low = 0;
    int high = integer_square_root(value);
    while (low <= high) {
        const int sum = low * low + high * high;
        if (sum == value) {
            first = low;
            second = high;
            return true;
        }
        if (sum < value) {
            ++low;
        } else {
            --high;
        }
    }
    return false;
}

// Tìm 3 số chính phương có tổng bằng value (giả thiết value không thuộc dạng
// cấm của Legendre). Duyệt third giảm dần, kiểm tra phần dư có tách được thành
// 2 số chính phương không.
std::tuple<int, int, int> three_squares(const int value) {
    for (int third = integer_square_root(value); third >= 0; --third) {
        int first = 0;
        int second = 0;
        if (two_squares(value - third * third, first, second)) {
            return {first, second, third};
        }
    }
    return {0, 0, 0};
}

int main() {
    std::ios::sync_with_stdio(false);
    std::cin.tie(nullptr);

    int tests;
    std::cin >> tests;
    while (tests-- > 0) {
        int value;
        std::cin >> value;
        if (value == 0) {
            std::cout << "0 0 0 0\n";
            continue;
        }

        // Bước 1 - khử hết thừa số 4: value = 4^k * reduced, scale = 2^k.
        // Nghiệm của reduced nhân lại với scale cho nghiệm của value.
        int reduced = value;
        int scale = 1;
        while (reduced % 4 == 0) {
            reduced /= 4;
            scale *= 2;
        }

        // Bước 2 - xử lý dạng cấm: nếu reduced ≡ 7 (mod 8) thì tách riêng
        // fourth = 1, phần còn lại reduced-1 chắc chắn là tổng 3 số chính phương.
        int fourth = 0;
        int target = reduced;
        if (reduced % 8 == 7) {
            target = reduced - 1;
            fourth = 1;
        }
        // Bước 3 - giải target bằng three_squares rồi nhân lại với scale.
        const auto [first, second, third] = three_squares(target);
        std::cout << first * scale << ' ' << second * scale << ' '
                  << third * scale << ' ' << fourth * scale << '\n';
    }
    return 0;
}
